### PR TITLE
Add logging of query IDs via QueryLogIterator class

### LINKED ISSUE #2305: QueryIterators/ScanIterators should log query id first
We need the query id logged in the Tserver logs as soon as the scan starts and again when the scan completes.  As a bonus it would be great if we could add the queryid to the thread name as well while it is executing.

### PR DESCRIPTION
**Pull request by SethSmucker**
- Create QueryLogIterator class that logs a query's ID to the TServer logs
- Add QueryLogIterator to DefaultQueryPlanner
- Add tserver.logging.active to QueryParameters, used to turn logging on and off
- Add tests for QueryLogIterator and TServerLogging

Resolves #2305

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/QueryParameters.java [hunk 1]
```diff
@@ -1,5 +1,7 @@
 package datawave.query;
 
+import datawave.query.iterator.QueryLogIterator;
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/QueryParameters.java:3]: This import is unused.

File: warehouse/query-core/src/main/java/datawave/query/iterator/QueryLogIterator.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+     * Wraps the init() method of the iterator above it, logging the start and end of the method along with its query id.
+     */
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+
+        try {
+            this.queryID = options.get(QUERY_ID);
+            this.source = source;
+            this.env = env;
+            logStartOf("init()");
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/iterator/QueryLogIterator.java:56]: Seems like this should be the first thing called in the function here. Also there is no delegation to the `source.init()`.

File: warehouse/query-core/src/main/java/datawave/query/iterator/QueryOptions.java
Comment (foster33) [inline warehouse/query-core/src/main/java/datawave/query/iterator/QueryOptions.java:1]: There are no actual changes here, so this class can probably be reverted.

File: warehouse/query-core/src/main/java/datawave/query/iterator/QueryLogIterator.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    private static final String CLASS_NAME = QueryLogIterator.class.getSimpleName();
+
+    private String queryID;
+    private SortedKeyValueIterator<Key,Value> source;
+    private IteratorEnvironment env;
+
+    /**
+     * Class constructor
+     */
+    public QueryLogIterator() {}
```
Comment (avgAGB) [inline warehouse/query-core/src/main/java/datawave/query/iterator/QueryLogIterator.java:35]: Could this be pruned out if it isn't being used?

File: warehouse/query-core/src/main/java/datawave/query/QueryParameters.java [hunk 4]
```diff
@@ -221,4 +223,9 @@ public class QueryParameters {
      * Used to specify model or DB fields that must be treated as strict (cannot be skipped if normalization fails)
      */
     public static final String STRICT_FIELDS = "strict.fields";
+
+    /**
+     * Controls whether a query's ID is logged on the tserver using {@link QueryLogIterator}
+     */
+    public static final String TSERVER_LOGGING_ACTIVE = "tserver.logging.active";
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/QueryParameters.java:230]: Please move this constant to the QueryOptions class instead.  QueryParameters are for parameters specified by users when submitting queries.  This is purly used as an option on the query log iterator.

File: warehouse/query-core/src/main/java/datawave/query/iterator/QueryLogIterator.java (hunk 2 above)
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/iterator/QueryLogIterator.java:56]: The tserver iterator setup will be calling the source.init with the options specific to that iterator.  No need to call source.init().

File: warehouse/query-core/src/main/java/datawave/query/QueryParameters.java (hunk 4 above)
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/QueryParameters.java:230]: You are setting this as a iterator option as well as pulling it out of the query parameters.  I don't believe the user submitting the query should be able to turn logging off and on.  That should be done when the query logic is being created (see QueryLogicFactory.xml).  You may want to add a line in there.  I think this constant should go away, and the default query planner should optionally setup the QueryLogIterator if that boolean is true in the configuration which it appears you are already doing.

File: warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java [hunk 5]
```diff
@@ -469,9 +477,21 @@ protected CloseableIterable<QueryData> process(ScannerFactory scannerFactory, Me
...
+                    if (logCfg == null) {
+                        logCfg = getQueryLogIterator(config, settings);
+                    }
+                    break;
+                }
+            }
         }
 
-        final QueryData queryData = new QueryData().withQuery(newQueryString).withSettings(Lists.newArrayList(cfg));
+        List<IteratorSetting> iteratorSettings = logCfg == null ? Collections.singletonList(cfg) : Lists.newArrayList(cfg, logCfg);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java:493]: We should only be adding the logCfg if config.isTserverLoggingActive().

Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java:493]: Optionally avoid creating the logCfg altogether if config.IsTserverLoggingActive() == false;.,

File: warehouse/query-core/src/test/java/datawave/query/config/ShardQueryConfigurationTest.java [hunk 6]
```diff
@@ -580,6 +580,9 @@ public void setUp() throws Exception {
         updatedValues.put("sortQueryBeforeGlobalIndex", true);
         defaultValues.put("sortQueryByCounts", false);
         updatedValues.put("sortQueryByCounts", true);
+
+        defaultValues.put("tserverLoggingActive", false);
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/config/ShardQueryConfigurationTest.java:584]: I think the default should be true

File: warehouse/query-core/src/test/java/datawave/query/tables/ShardQueryLogicTest.java [hunk 7]
```diff
@@ -360,4 +360,9 @@ public void testNegativeFilterRegexV2() throws Exception {
 
         runTestQuery(expected, queryString, format.parse("20091231"), format.parse("20150101"), extraParameters);
     }
+
+    @Test
+    public void testTserverLoggingActive() {
+
```
Comment (ivakegg) [inline warehouse/query-core/src/test/java/datawave/query/tables/ShardQueryLogicTest.java:366]: This is not much of a test....

**Review by ivakegg — CHANGES_REQUESTED** [review]
See comments

**Comment by SethSmucker** [comment]
Superseded by #2428, all changes suggested have been added there. That PR pulls directly from my fork instead of using a middle-man branch between my fork's branch and upstream/integration.

**Pull request closed without merging**